client/src/views/TableDetail.vue

```diff
@@ -88,34 +88,38 @@
     </v-content>
   </v-container>
 </template>
-<script>
-import api from '@/api'
 
-export default {
+<script lang="ts">
+import Vue from 'vue';
+
+import api from '@/api';
+import { KeyValue, TableRow } from '@/types';
+
+export default Vue.extend({
   name: 'TableDetail',
   props: ['workspace', 'table'],
-  data () {
+  data() {
     return {
-      rowKeys:[],
-      headers:[],
+      rowKeys: [] as KeyValue[][],
+      headers: [] as Array<keyof TableRow>,
```

Comment:
So, I went back and forth on this a bit. I originally had code that declared const default values for each field (typed with the usual syntax), then returned the object version, like this:

```typescript
function foobar() {
  const x: number = 10;
  const y: KeyValue[][] = [];

  return {
    x,
    y,
  };
}
```

In my opinion, this is hard to read. The other `as`-free way to do this is not much better:

```typescript
function foobar() : { x: number, y: KeyValue[][] } {
  return {
    x: 10,
    y: [],
  };
}
```

Actually, I think it's quite a bit worse. The approach I adopted has some advantages:
1. It is more readable that either choice above.
2. It doesn't repeat the keys of the final object.

The admonition in that article not to use `as` seems a bit defective. If the advice is to avoid casting values to `any` just to satisfy the compiler, then I agree 100%. However, `as` has its uses: for instance, if a value `x` has type `A | B`, and some runtime test tells you that your value actually has type `A`, then writing `(x as A)` in order to call a method of `A` on it is perfectly valid.

In this case, I'm using it idiomatically to cast an ambiguous value to a value of a concrete type. That is, `[]` is an empty list in every list type; I'm using `as` to specify which list type I'm actually using. I would argue this particular use of `as` is a good one, since it reads clearly, its semantics are not being abused, and it saves keystrokes.

Reply:
Investigating a bit further, "as", along with the const declarations and cast operator, are put forward as three possible solutions. No one seems to suggest that "as" in this instance is bad, so I think your current approach is reasonable.

https://stackoverflow.com/a/12832404

There are also some interesting suggestions to typescript itself that might have a better solution in the future, though I don't fully understand the discussion:

https://github.com/microsoft/TypeScript/issues/7481#issuecomment-250297436